bigtable-hbase-dataflow/src/main/java/com/google/cloud/bigtable/dataflow/CloudBigtableConnectionPool.java

```diff
@@ -0,0 +1,120 @@
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.LinkedHashMultimap;
+
+/**
+ * Pubsub and other windowed sources can have a large quantity of bundles in short amounts of time.
+ * {@link CloudBigtableIO.AbstractCloudBigtableTableWriteFn} should not create a connection per
+ * bundle, since that could happen ever few milliseconds. Rather, it should rely on a connection
+ * pool to better manage connection life-cycles.
+ */
+public class CloudBigtableConnectionPool {
+
+  private static final int MAX_TTL_MILLISECONDS = 30 * 60 * 1000;
+
+  /**
+   * This class abstracts a connection that could be returned to a
+   * {@link CloudBigtableConnectionPool}.
+   */
+  public static class PoolEntry {
+    private final String key;
+    private final Connection connection;
+    private final long expiresTimeMs;
+
+    private static long getExpiration() {
+      return System.currentTimeMillis() + MAX_TTL_MILLISECONDS;
+    }
+
+    public PoolEntry(String key, Connection connection) {
+      this(key, connection, getExpiration());
+    }
+
+    @VisibleForTesting
+    PoolEntry(String key, Connection connection, long expiresTimeMs) {
+      this.key = key;
+      this.connection = connection;
+      this.expiresTimeMs = expiresTimeMs;
+    }
+
+    public Connection getConnection() {
+      return connection;
+    }
+
+    public String getKey() {
+      return key;
+    }
+
+    public boolean isExpired() {
+      return System.currentTimeMillis() > expiresTimeMs;
+    }
+    
+    @Override
+    public String toString() {
+      return String.format("%s key=%s, hash=%s", getClass().getName(), key, super.toString());
+    }
+  }
+
+  private final LinkedHashMultimap<String, PoolEntry> connections = LinkedHashMultimap.create();
+
+  public PoolEntry getConnection(Configuration config) throws IOException {
+    String key = BigtableOptionsFactory.fromConfiguration(config).getClusterName().toString();
+    return getConnection(config, key);
+  }
+
+  @VisibleForTesting
+  protected synchronized PoolEntry getConnection(Configuration config, String key) throws IOException {
+    Set<PoolEntry> entries = connections.get(key);
+    if (entries.isEmpty()) {
+      return createConnection(config, key);
+    }
+    Iterator<PoolEntry> iterator = entries.iterator();
+    while (iterator.hasNext()) {
+      PoolEntry entry = iterator.next();
+      iterator.remove();
+      if (entry.isExpired()) {
+        entry.connection.close();
+      } else {
+        return entry;
+      }
+    }
+    return createConnection(config, key);
+  }
+
+  @VisibleForTesting
+  protected PoolEntry createConnection(Configuration config, String key) throws IOException {
+    return new PoolEntry(key, new BigtableConnection(config));
+  }
+
+  public synchronized void returnConnection(PoolEntry entry) throws IOException {
+    if (entry.isExpired()) {
+      entry.connection.close();
+    } else {
+      connections.put(entry.getKey(), entry);
```

Comment:
yes.  The idea is that getConnection() will create a connection, if none are in the pool.  After the PTransform is done, it will return the new connection to the pool to be reused.